bigtable_input.py

```diff
@@ -169,24 +173,98 @@ def create(self):
         and their properties.
         """
         if self.bt_table.exists():
-            print('Table already exists')
+            utils.dbg('Table already exists')
             return
 
         max_versions_rule = column_family.MaxVersionsGCRule(1)
         self.bt_table.create(column_families={
             METADATA: max_versions_rule,
             TFEXAMPLE: max_versions_rule})
 
+    @property
     def latest_game_number(self):
         """Return the number of the next game to be written."""
-        game_counter = b'game_counter'
         table_state = self.bt_table.read_row(
             TABLE_STATE,
             filter_=row_filters.ColumnRangeFilter(
-                METADATA, game_counter, game_counter))
+                METADATA, GAME_COUNTER, GAME_COUNTER))
         if table_state is None:
             return 0
-        return cbt_intvalue(table_state.cell_value(METADATA, game_counter))
+        return cbt_intvalue(table_state.cell_value(METADATA, GAME_COUNTER))
+
+    @latest_game_number.setter
+    def latest_game_number(self, latest):
+        table_state = self.bt_table.row(TABLE_STATE)
+        table_state.set_cell(METADATA, GAME_COUNTER, int(latest))
+        table_state.commit()
+
+    def games_by_time(self, start_game, end_game):
+        """Given a range of games, return the games sorted by timestamp.
+
+        Returns [(timestamp, game), ...]
+        """
+        move_count = b'move_count'
+        rows = self.bt_table.read_rows(
+            ROWCOUNT_PREFIX.format(start_game),
+            ROWCOUNT_PREFIX.format(end_game),
+            filter_=row_filters.ColumnRangeFilter(
+                METADATA, move_count, move_count))
+        def parse(r):
+            rk = str(r.row_key, 'utf-8')
+            game = _game_from_counter.match(rk).groups()[0]
+            return (r.cells[METADATA][move_count][0].timestamp, game)
+        return sorted([parse(r) for r in rows], key=operator.itemgetter(0))
+
+    def delete_row_range(self, format_str, start_game, end_game):
+        row_keys = make_single_array(
+            self.tf_table.keys_by_range_dataset(
+                format_str.format(start_game),
+                format_str.format(end_game)))
+        row_keys = list(row_keys)
+        utils.dbg('Deleting %d rows:  %s..%s' % (
+            len(row_keys), row_keys[0], row_keys[-1]))
+        rows = [self.bt_table.row(k) for k in row_keys]
+        batch = []
+        def mutate_batch():
+            nonlocal batch
+            utils.dbg('...deleting batch %s..%s' % (
+                batch[0].row_key, batch[-1].row_key))
+            self.bt_table.mutate_rows(batch)
+            batch = []
+        while rows:
+            r = rows.pop()
+            r.delete()
+            batch.append(r)
+            if len(batch) > 50000:
```

Comment:
what's the 50k?

Reply:
Oops! That's half the maximum Bigtable mutation batch size. There is a module constant for this, and I'll use it. I was using half the batch size for testing.